File: matiasmasso/Mat.NET/DT2/DTONavbar.vb
Public Class DTONavbar
    Inherits DTONavbarItem
    Property format As formats

    Public Enum Formats
        horizontal
        vertical
    End Enum

    Shared Shadows Function Factory(oFormat As DTONavbar.Formats, Optional sId As String = "") As DTONavbar
        Dim retval As New DTONavbar
        With retval
            .Id = sId
            .format = oFormat
            .Items = New List(Of DTONavbarItem)
        End With
        Return retval
    End Function

    Public Shadows Function Html() As String
        Dim navAttributes As New System.Text.StringBuilder
        navAttributes.AppendLine(" class='" & Me.format.ToString & "'")
        If MyBase.Id > "" Then
            navAttributes.AppendLine(" id='" & MyBase.Id & "'")
        End If

        Dim sb As New System.Text.StringBuilder
        sb.AppendLine("<nav" & navAttributes.ToString & ">")
        sb.Append("<ul>")
        For Each oItem As DTONavbarItem In MyBase.Items
            sb.Append(oItem.Html())
        Next
        sb.AppendLine("</ul>")
        sb.AppendLine("</nav>")
        Dim retval As String = sb.ToString
        Return retval
    End Function
End Class

Public Class DTONavbarItem
    Property Id As String
    Property Title As String
    Property NavigateTo As String
    Property Tooltip As String
    Property Selected As Boolean

    Property Items As List(Of DTONavbarItem)

    Public Sub New()
        MyBase.New
        _Items = New List(Of DTONavbarItem)
    End Sub

    Shared Function Factory(sTitle As String, Optional sNavigateTo As String = "#", Optional BlSelected As Boolean = False) As DTONavbarItem
        Dim retval As New DTONavbarItem
        With retval
            .Title = sTitle
            .NavigateTo = sNavigateTo
            .Selected = BlSelected
            .Items = New List(Of DTONavbarItem)
        End With
        Return retval
    End Function

    Public Function AddItem(sTitle As String, Optional sNavigateTo As String = "#", Optional BlSelected As Boolean = False) As DTONavbarItem
        Dim retval = DTONavbarItem.Factory(sTitle, sNavigateTo, BlSelected)
        _Items.Add(retval)
        Return retval
    End Function

    Public Function Html() As String
        Dim sb As New System.Text.StringBuilder
        sb.Append("<li")
        sb.Append(If(_Selected, " class='selected'", ""))
        sb.Append(If(_Id > "", " id='" & _Id & "'", ""))
        sb.Append(">")
        sb.Append("<a href='" & _NavigateTo & "' ")
        sb.Append(If(_Id > "", " data-ref='" & _Id & "'", ""))
        sb.Append(">")
        sb.Append(_Title)
        sb.Append("</a>")

        If _Items.Count > 0 Then
            sb.Append("<ul>")
            For Each oItem As DTONavbarItem In _Items
                sb.Append(oItem.Html())
            Next
            sb.AppendLine("</ul>")
        End If
        sb.Append("</li>")
        Dim retval As String = sb.ToString
        Return retval
    End Function
End Class

File: matiasmasso/Mat.NET/DTO/DTONavbar.cs
using System.Collections.Generic;

namespace DTO
{
    public class DTONavbar : DTONavbarItem
    {
        public Formats Format { get; set; }

        public enum Formats
        {
            horizontal,
            vertical
        }

        public static DTONavbar Factory(DTONavbar.Formats oFormat, string sId = "")
        {
            DTONavbar retval = new DTONavbar();
            {
                var withBlock = retval;
                withBlock.Id = sId;
                withBlock.Format = oFormat;
                withBlock.Items = new List<DTONavbarItem>();
            }
            return retval;
        }

        public new string Html()
        {
            System.Text.StringBuilder navAttributes = new System.Text.StringBuilder();
            navAttributes.AppendLine(" class='" + this.Format.ToString() + "'");
            if (!string.IsNullOrEmpty(base.Id))
                navAttributes.AppendLine(" id='" + base.Id + "'");

            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.AppendLine("<nav" + navAttributes.ToString() + ">");
            sb.Append("<ul>");
            foreach (DTONavbarItem oItem in base.Items)
                sb.Append(oItem.Html());
            sb.AppendLine("</ul>");
            sb.AppendLine("</nav>");
            string retval = sb.ToString();
            return retval;
        }
    }

    public class DTONavbarItem
    {
        public string Id { get; set; }
        public string Title { get; set; }
        public string NavigateTo { get; set; }
        public string Tooltip { get; set; }
        public bool Selected { get; set; }

        public List<DTONavbarItem> Items { get; set; }

        public DTONavbarItem() : base()
        {
            Items = new List<DTONavbarItem>();
        }

        public static DTONavbarItem Factory(string sTitle, string sNavigateTo = "#", bool BlSelected = false)
        {
            DTONavbarItem retval = new DTONavbarItem();
            {
                var withBlock = retval;
                withBlock.Title = sTitle;
                withBlock.NavigateTo = sNavigateTo;
                withBlock.Selected = BlSelected;
                withBlock.Items = new List<DTONavbarItem>();
            }
            return retval;
        }

        public DTONavbarItem AddItem(string sTitle, string sNavigateTo = "#", bool BlSelected = false)
        {
            var retval = DTONavbarItem.Factory(sTitle, sNavigateTo, BlSelected);
            Items.Add(retval);
            return retval;
        }

        public string Html()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append("<li");
            sb.Append(Selected ? " class='selected'" : "");
            sb.Append(string.IsNullOrEmpty(Id) ? "" : " id='" + Id + "'");
            sb.Append(">");
            sb.Append("<a href='" + NavigateTo + "' ");
            sb.Append(string.IsNullOrEmpty(Id) ? "" : "data-ref='" + Id + "'");
            sb.Append(">");
            sb.Append(Title);
            sb.Append("</a>");

            if (Items.Count > 0)
            {
                sb.Append("<ul>");
                foreach (DTONavbarItem oItem in Items)
                    sb.Append(oItem.Html());
                sb.AppendLine("</ul>");
            }
            sb.Append("</li>");
            string retval = sb.ToString();
            return retval;
        }
    }
}
